cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -2392,6 +2392,29 @@ struct NonZeroVisitor {
 
     return Status::OK();
   }
+
+  template <typename Type>
+  enable_if_t<is_decimal128_type<Type>::value || is_decimal256_type<Type>::value, Status>
+  Visit(const Type&) {
```

Comment:
This is morally the same code as the previous override above, can we merge them in a single method declaration?

Reply:
I have updated this, and added a specialisation for bool, `Status Visit(const BooleanType&)`.

Reply:
I simplified this more and removed the bool specialization which wasn't actually necessary.